runtime/trampolines.m

```diff
@@ -721,57 +723,119 @@
 
+struct gchandle_dictionary_entry {
+	GCHandle gc_handle;
+	enum XamarinGCHandleFlags flags;
+};
+
+static void
+release_gchandle_dictionary_entry (CFAllocatorRef allocator, const void *value)
+{
+	free ((void *) value);
+}
+
 static const char *associated_key = "x"; // the string value doesn't matter, only the pointer value.
 void
-xamarin_set_gchandle_trampoline (id self, SEL sel, uint32_t gc_handle)
+xamarin_set_gchandle_trampoline (id self, SEL sel, GCHandle gc_handle, enum XamarinGCHandleFlags flags)
 {
-	// COOP: Called by ObjC (when the setGCHandle: selector is called on an object).
+	// COOP: Called by ObjC (when the setGCHandle:flags: selector is called on an object).
 	// COOP: Safe mode upon entry, and doesn't access managed memory, so no need to change.
 	MONO_ASSERT_GC_SAFE;
 	
 	/* This is for types registered using the dynamic registrar */
 	XamarinAssociatedObject *obj;
 	obj = objc_getAssociatedObject (self, associated_key);
-	if (obj == NULL && gc_handle != 0) {
+	if (obj == NULL && gc_handle != INVALID_GCHANDLE) {
 		obj = [[XamarinAssociatedObject alloc] init];
 		obj->gc_handle = gc_handle;
+		obj->flags = flags;
 		obj->native_object = self;
 		objc_setAssociatedObject (self, associated_key, obj, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
 		[obj release];
 	}
 
-	if (obj != NULL)
+	if (obj != NULL) {
 		obj->gc_handle = gc_handle;
+		obj->flags = flags;
+	}
 	
 	pthread_mutex_lock (&gchandle_hash_lock);
-	if (gchandle_hash == NULL)
+	if (gchandle_hash == NULL) {
+		CFDictionaryValueCallBacks value_callbacks;
+		value_callbacks.release = release_gchandle_dictionary_entry;
 		gchandle_hash = CFDictionaryCreateMutable (kCFAllocatorDefault, 0, NULL, NULL);
-	if (gc_handle == 0) {
+	}
+	if (gc_handle == INVALID_GCHANDLE) {
 		CFDictionaryRemoveValue (gchandle_hash, self);
 	} else {
-		CFDictionarySetValue (gchandle_hash, self, GINT_TO_POINTER (gc_handle));
+		struct gchandle_dictionary_entry *entry = (struct gchandle_dictionary_entry *) malloc (sizeof (struct gchandle_dictionary_entry));
+		entry->gc_handle = gc_handle;
+		entry->flags = flags;
+		CFDictionarySetValue (gchandle_hash, self, entry);
 	}
 	pthread_mutex_unlock (&gchandle_hash_lock);
 }
 
-uint32_t
+GCHandle
 xamarin_get_gchandle_trampoline (id self, SEL sel)
 {
 	// COOP: Called by ObjC (when the getGCHandle selector is called on an object).
 	// COOP: Safe mode upon entry, and doesn't access managed memory, so no need to switch.
 	MONO_ASSERT_GC_SAFE;
 	
 	/* This is for types registered using the dynamic registrar */
-	uint32_t gc_handle = 0;
+	GCHandle gc_handle = INVALID_GCHANDLE;
 	pthread_mutex_lock (&gchandle_hash_lock);
-	if (gchandle_hash != NULL)
-		gc_handle = GPOINTER_TO_UINT (CFDictionaryGetValue (gchandle_hash, self));
+	if (gchandle_hash != NULL) {
+		struct gchandle_dictionary_entry *entry;
+		entry = (struct gchandle_dictionary_entry *) CFDictionaryGetValue (gchandle_hash, self);
+		if (entry != NULL)
+			gc_handle = entry->gc_handle;
```

Comment:
Questions, and I know that you are going to tell me we don't care, but I want to be sure, does the value gchandle_hash can contain NULL, if that is the case, should we use [CFDictionaryGetValueIfPresent](https://developer.apple.com/documentation/corefoundation/1516739-cfdictionarygetvalueifpresent?language=objc) instead?

Reply:
```
	if (gchandle_hash != NULL) {
```
a few lines up

Reply:
The dictionary shouldn't contain NULL values, but even if it does, this is the right way to handle them, so yeah, we don't care 😄